pyomo/contrib/piecewise/triangulations.py

```diff
@@ -0,0 +1,717 @@
+from enum import Enum
+from pyomo.common.errors import DeveloperError
+from pyomo.common.dependencies import numpy as np
+from pyomo.contrib.piecewise.ordered_3d_j1_triangulation_data import (
+    hamiltonian_paths as incremental_3d_simplex_pair_to_path,
+)
+
+
+class Triangulation(Enum):
+    Unknown = 0
+    AssumeValid = 1
+    Delaunay = 2
+    J1 = 3
+    OrderedJ1 = 4
+
+
+# Get an unordered J1 triangulation, as described by [1], of a finite grid of
+# points in R^n having the same odd number of points along each axis.
+# References
+# ----------
+# [1] J.P. Vielma, S. Ahmed, and G. Nemhauser, "Mixed-integer models
+#     for nonseparable piecewise-linear optimization: unifying framework
+#     and extensions," Operations Research, vol. 58, no. 2, pp. 305-315,
+#     2010.
+def get_unordered_j1_triangulation(points, dimension):
+    points_map, num_pts = _process_points_j1(points, dimension)
+    simplices_list = _get_j1_triangulation(points_map, num_pts - 1, dimension)
+    return _FakeScipyTriangulation(
+        points=np.array(points),
+        simplices=np.array(simplices_list),
+        coplanar=np.array([]),
+    )
+
+
+# Get an ordered J1 triangulation, according to [1], with the additional condition
+# added from [2] that simplex vertices are also ordered such that the final vertex
+# of each simplex is the first vertex of the next simplex.
+# References
+# ----------
+# [1] Michael J. Todd. "Hamiltonian triangulations of Rn". In: Functional
+#     Differential Equations and Approximation of Fixed Points. Ed. by
+#     Heinz-Otto Peitgen and Hans-Otto Walther. Berlin, Heidelberg: Springer
+#     Berlin Heidelberg, 1979, pp. 470–483. ISBN: 978-3-540-35129-0.
+# [2] J.P. Vielma, S. Ahmed, and G. Nemhauser, "Mixed-integer models
+#     for nonseparable piecewise-linear optimization: unifying framework
+#     and extensions," Operations Research, vol. 58, no. 2, pp. 305-315,
+#     2010.
+def get_ordered_j1_triangulation(points, dimension):
+    points_map, num_pts = _process_points_j1(points, dimension)
+    if dimension == 2:
+        simplices_list = _get_ordered_j1_triangulation_2d(points_map, num_pts - 1)
+    elif dimension == 3:
+        simplices_list = _get_ordered_j1_triangulation_3d(points_map, num_pts - 1)
+    else:
+        simplices_list = _get_ordered_j1_triangulation_4d_and_above(
+            points_map, num_pts - 1, dimension
+        )
+    return _FakeScipyTriangulation(
+        points=np.array(points),
+        simplices=np.array(simplices_list),
+        coplanar=np.array([]),
+    )
+
+
+# Duck-typed thing that looks reasonably similar to an instance of scipy.spatial.Delaunay
+# Fields:
+#   - points: list of P points as P x n array
+#   - simplices: list of M simplices as P x (n + 1) array of point _indices_
+#   - coplanar: list of N points omitted from triangulation as tuples of (point index,
+#     nearest simplex index, nearest vertex index), stacked into an N x 3 array
+class _FakeScipyTriangulation:
```

Comment:
You could probably just call this `_Triangulation` or something--maybe someday we'll make it do more triangulation-y things if we need to.

Reply:
Changed